### ISSUE TITLE
Oracle VM does not delete the datafiles

### ISSUE THREAD

**Issue by sorin-pop** [djavorszky/ddn#209]
While investigating #207, I noticed that the physical datafiles are not deleted on the file system. The user is dropped, its tablespace is dropped, the datafile objects belonging to the tablespce are dropped, but the actual files on the file system remain there. This is a major problem and very worrying.

After an initial research, it seems like this is a known issue on Windows platforms: https://dba.stackexchange.com/questions/143205/cant-delete-oracle-datafiles-after-drop-tablespace-including-content-and-datafi

https://community.oracle.com/thread/867026?db=5

I am further looking into this. (we might need to think about rather using Linux, but that's also kind of problematic: having it set up with IT, installing Oracle on it, which is not an easy task with Linux
)

**Comment by djavorszky** [comment]
That's interesting.. Could we delete the files manually? E.g. grab the location of the datafiles and simply call `os.Remove(filename)` on them manually?

Also, note #171, that could be a logical next step.

**Comment by sorin-pop** [comment]
Yes, I tried deleting manually a few datafiles and it worked, so we could programmatically call os.Remove from the agent as a fallback (although doesn't sound like a very elegant thing to do. Basically, DROP TABLASPACE INCLUDING CONTENTS AND DATAFILES should do this, at least that's what I thought until now). However, the first post I linked above seems to suggest that the stuck datafiles are not always deletable manually because Oracle/Windows holds them. 

I wouldn't mind if #171 would be set in motion. It has just been sitting there, but now...we may need to do it.

But I'm still looking into this problem anyway...will update soon.

**Comment by sorin-pop** [comment]
Seems like it's an Oracle bug that comes on Windows: https://community.oracle.com/thread/867026?db=5

> Normally, "drop tablespace ... including contents and datafiles" should remove the datafiles. On Windows there are some bugs causing this error and not deleting the datafiles. See bugs 2871967 and 2338968 if you have a Metalink account.

**Comment by sorin-pop** [comment]
I could not find much more info on the subject. As I don't have an offcial Oracle support metalink account, I could not see the alleged official pages related to this problem.

So in summary:
- based on forum comments, it seems to be an Oracle+Windows bug, which might have even been fixed already, and it looks that we happen to use such an Oracle+Windows version combination that has this problem.
- based on forum comments, there's really no other workaround than deleting the files manually, after a **database sever restart** (which we do not want to do, of course)

For now, I can periodically check this and manually delete the orphaned datafiles. (hopefully it will work like today, without having to stop the db server)
On the long run, I think we should create another platform: another server (preferably Linux then...), or a different Oracle server version (we use 12.1 now, maybe 12.2 does not have this problem...)

What do you think @djavorszky @balcsida ?

**Comment by djavorszky** [comment]
Thanks for looking into it @sorin-pop! As we're planning on moving to the official docker container per #171, we shouldn't invest too much effort into fixing this. How hard it is to manually find which datafiles are orphaned and delete them? If it's too much effort (e.g. requires manual cross reference or something like that), then we should at least save which datafiles should be deleted when dropping the user/schema/datafile within Oracle.

What do you think?

**Comment by sorin-pop** [comment]
It's not that hard to do this check manually. I could even automate it with a PL/SQL routine(I was just thinking about this last night), although it's not that important: it's really not difficult to manually identify the orphan datafiles. But deleting them might be tricky: we might have to stop the db server in order to do that. I think I did not have to do it last time, because the server (VM) was restarted once since these datafiles were created, when I tested the automatic restart of the CloudDB agent.

So I am really looking forward to #171.
